src/Language/Spectacle/Syntax/Modal.hs

```diff
@@ -0,0 +1,165 @@
+
+-- ---------------------------------------------------------------------------------------------------------------------
+
+-- | The modal operator 'always' qualifies a formula @p@ such that for all actions taken in a specification, the formula
+-- @p@ must be true.
+--
+-- [Valid]
+--
+-- @
+-- ... ───── p ───── p ───── p ───── p ───── p ───── ...
+-- @
+--
+-- [Invalid]
+--
+-- @
+-- ... ───── p ───── p ───── p ───── ○ ───── p ───── ...
+-- @
+--
+-- @since 0.1.0.0
+always :: Member Modal effs => Lang ctx effs Bool -> Lang ctx effs Bool
+always m = scope (Always m)
+{-# INLINE always #-}
+
+-- | The modal operator 'eventually' qualifies a formula @p@ such that @p@ must be true /now/, or some time in the
+-- future. 'eventually' is equivalent to 'upUntil' with a trivial formula on the left-hand side:
+--
+-- @
+-- 'eventually' p = 'upUntil' ('pure' 'True') p
+-- @
+--
+-- [Valid]
+--
+-- @
+-- ... ───── ○ ───── ○ ───── ○ ───── p ───── ○ ───── ...
+--
+-- ... ───── ○ ───── ○ ───── p ───── ○ ───── p ───── ...
+--
+-- ... ───── p ───── p ───── p ───── p ───── p ───── ...
+-- @
+--
+-- [Invalid]
+--
+-- @
+-- ... ───── ○ ───── ○ ───── ○ ───── ○ ───── ○ ───── ...
+-- @
+--
+-- @since 0.1.0.0
+eventually :: Member Modal effs => Lang ctx effs Bool -> Lang ctx effs Bool
+eventually = upUntil (pure True)
+
+-- | The modal operator 'upUntil' (strong) says that for some formula @p `upUntil` q@, @p@ must be true up until the
+-- point where @q@ is true, and that @q@ /must/ be true some time in the future.
+--
+-- [Valid]
+--
+-- @
+-- ... ───── p ───── q ───── q ───── ○ ───── ○ ───── ...
+--
+-- ... ─── p ∧ q ─── ○ ───── ○ ───── ○ ───── ○ ───── ...
+--
+-- ... ───── p ───── q ───── ○ ───── p ───── q ───── ...
+-- @
+--
+-- [Invalid]
+--
+-- @
+-- ... ───── p ───── p ───── p ───── ○ ───── q ───── ...
+--
+-- ... ───── p ───── p ───── p ───── p ───── p ───── ...
+--
+-- ... ───── ○ ───── ○ ───── ○ ───── ○ ───── ○ ───── ...
+-- @
+--
+-- @since 0.1.0.0
+upUntil :: Member Modal effs => Lang ctx effs Bool -> Lang ctx effs Bool -> Lang ctx effs Bool
+upUntil m n = scope (UpUntil m n)
+
+-- | Introduces a prime effect underneath the 'Logic' and 'Modal' effects in a temporal formula. The prime effect is
+-- used to replace the plain variable usage on the right-hand side of 'upUntil' with prime variables. This is so we can
+-- substitute the new values obtained by running an action to check if a 'upUntil' is satisfied.
+--
+-- 'Prime' is introduced only when needed to avoid giving users access to the 'prime' syntax in temporal formula.
+--
+-- @since 0.1.0.0
+introducePrimes :: Lang ctx (Modal ': Logic ': effs) a -> Lang ctx (Modal ': Logic ': Prime ': effs) a
+introducePrimes = \case
+  Pure x -> pure x
+  Op op k -> Op (extendOp op) (introducePrimes . k)
+  Scoped scoped loom -> Scoped (extendS scoped) (loom ~>~ hoist introducePrimes)
+  where
+    extendOp :: Op (Modal ': Logic ': effs) a -> Op (Modal ': Logic ': Prime ': effs) a
+    extendOp = \case
```

Comment:
Since you're only adapting one case, you can let the others fall through:

```suggestion
    extendOp = \case
      OThere (OThere op) -> OThere (OThere (OThere op))
      op -> op
```

Reply:
`Op` is witnessing `(Modal ': Logic ': effs)` here so you have to peel off each `OThere` up to `effs` and rebuild it with the same constructor this time witnessing a different list in the type parameter. So for example `OHere op :: Op (Modal ': Logic ': effs)` is being sent to `OHere op :: Op (Modal ': Logic ': Prime ': effs)`. Otherwise GHC thinks you're constructing an infinite type.

Reply:
Oh, I wasn't paying attention to the type.